syntax = "proto3";

package rendezvous_server;

service ServerService {
  rpc RegisterRequest(RegisterRequestMessage) returns (Empty);
  rpc RegisterBranch(RegisterBranchMessage) returns (Empty);
  rpc CloseBranch(CloseBranchMessage) returns (Empty);
}

/* Helpers */
message Empty {

}

/* Metadata for server replication */
message RequestContext {
  string prev_service = 1;
  map<string, int32> versions = 2;
}

/* Register Request */
message RegisterRequestMessage {
  string rid = 1;
}

/* Registers Branches */
message RegisterBranchMessage {
  string root_rid = 1;
  string sub_rid = 2;
  string core_bid = 3;
  string service = 4;
  string tag = 5;
  repeated string regions = 6;
  bool monitor = 7;
  RequestContext context = 8;
}

/* Close Branch */
message CloseBranchMessage {
  string root_rid = 1;
  string sub_rid = 2;
  string core_bid = 3;
  string region = 4;
  RequestContext context = 5;
}